### PR TITLE
Catalog: Show install error with incompatible version

### LINKED ISSUE #55806: Plugin Catalog: Backend plugin install version incompatibility
**What happened**:
image

**What you expected to happen**:
For it to be clear that the plugin is not compatible (and therefore cannot install or you can try to install the plugin but at your own risk)

IE re-use this in the UI like we do when checking directly with the `grafanaDependecy` plugin.json field 
image

**How to reproduce it (as minimally and precisely as possible)**:
Find a backend plugin with no matching OS + architecture for your system

**Anything else we need to know?**:

**Environment**:
- Grafana version:
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by andresmgot**
**What is this feature?**

This PR contains 3 minor changes:

 - If a plugin installation fail, don't show the "success" message.
 - Render error information in the installation page.
 - Disable the installation button if it failed.

Before:

image

After:

Screenshot from 2023-04-14 16-22-40

**Why do we need this feature?**

Better UX when trying to install incompatible plugins

**Who is this feature for?**

Users

**Which issue(s) does this PR fix?**:

Fixes #55806

**Special notes for your reviewer**:

Ideally, we could proactively show a warning if the plugin is incompatible before clicking install but:

 - We don't have the backend architecture info in the frontend (which is what causes the error in the screenshot).
 - There could be other possible errors in the backend but the frontend does not need to be aware of everything.

### CODE REVIEW

**Review by leventebalogh — COMMENTED** [review]
Nice to see this being fixed! 🙏 

A quick idea: what if instead of showing the error alert inline, we would show an error notification with the message that comes down the pipe and would not show the default warning notification? (We could still disable the install button.)

```javascript
// We could do something like this
if (isFetchError(e)) {
    appEvents.emit(AppEvents.alertError, [e.data]);
    return thunkApi.rejectWithValue(e.data);
}
```

**Comment by andresmgot** [comment]
> @andresmgot what exactly is missing here? Is it just the Grafana version?

No, what we need in this case is the server OS + arch (e.g. linux-amd64) but even if we could add that, I don't think it's the job of the frontend to foresee all the possible backend validations or errors.

**Comment by leventebalogh** [comment]
Hmm, yea, that's not ideal indeed. (Another idea that came to my mind would be to change the text on the install button and disable it, although that could look bad with a too long error message.)

In case we keep the inline error, I think we should get rid of the warning notification as it feels a bit redundant, wdyt? (Using `showErrorAlert: false`).

It could also make sense to involve @grafana/ux, maybe they have more ideas.

**Comment by amy-super** [comment]
👋 this seems like a great candidate for a UX Feedback Session. You can find details pinned to the top of `#ux` for how to sign up.

**Comment by andresmgot** [comment]
> wave this seems like a great candidate for a UX Feedback Session. You can find details pinned to the top of `#ux` for how to sign up.

will do :+1:

**Review by leventebalogh — APPROVED** [review]
Code wise LGTM, let's see what comes out of the UX meeting (although we can tackle that in a separate PR as well I think).

**Comment by andresmgot** [comment]
a summary from the UX session:
 - Keep the static error (inline) vs the temporary banner
 - [ ] Move the error Alert below the Install button to avoid its jump
 - [ ] Improve the error message to give some actionable item or more specific error details

**Review by wbrowne — COMMENTED** [review]
When testing I don't see what the PR description shows, all I see is this failure message and the greyed out `Install` button. Is that correct?

image

**Comment by andresmgot** [comment]
> When testing I don't see what the PR description shows, all I see is this failure message and the greyed out `Install` button. Is that correct?

It depends on how you are triggering the error. I am using `error.message` as the title and `error.error` as the body of the error. This is the format I've seen it's set when the backend returns an error. How did you generate it?

**Comment by wbrowne** [comment]
> > When testing I don't see what the PR description shows, all I see is this failure message and the greyed out `Install` button. Is that correct?
> 
> It depends on how you are triggering the error. I am using `error.message` as the title and `error.error` as the body of the error. This is the format I've seen it's set when the backend returns an error. How did you generate it?

I just hardcoded my OS to be Windows ([here](https://github.com/grafana/grafana/blob/main/pkg/api/plugins.go#L449)) so that GCOM would never find the `Windows`+`amd64` combo and that was the result!

**Comment by andresmgot** [comment]
> I just hardcoded my OS to be Windows ([here](https://github.com/grafana/grafana/blob/main/pkg/api/plugins.go#L449)) so that GCOM would never find the `Windows`+`amd64` combo and that was the result!

I get a proper error if I do the same @wbrowne:

Screenshot from 2023-04-12 10-02-33

I am not sure what you mean with GCOM though. Also, have you rebuilt both backend and frontend? (I double checked pulling the latest changes in main).

**Comment by wbrowne** [comment]
> > I just hardcoded my OS to be Windows ([here](https://github.com/grafana/grafana/blob/main/pkg/api/plugins.go#L449)) so that GCOM would never find the `Windows`+`amd64` combo and that was the result!
> 
> I get a proper error if I do the same @wbrowne:
> 
> Screenshot from 2023-04-12 10-02-33
> 
> I am not sure what you mean with GCOM though. Also, have you rebuilt both backend and frontend? (I double checked pulling the latest changes in main).

Ah I see the problem - I was running Enterprise! Looks like that changes how we display the error

**Comment by andresmgot** [comment]
> Ah I see the problem - I was running Enterprise! Looks like that changes how we display the error

Sorry but I still don't see that affecting that behavior (it shouldn't), this is an enterprise plugin:

Screenshot from 2023-04-13 11-44-45

Maybe we can jump into a call and debug what may be happening on your side?

**Comment by wbrowne** [comment]
@andresmgot Weird.. Okay let me give it one more shot

**Comment by wbrowne** [comment]
@andresmgot Okay I found the culprit - I had the mode as `production`, not `development`. Looks like we don't show those error messages for prod

**Comment by xnyo** [comment]
Ah makes sense! -- I checked and the error is logged to backend stdout when running in production mode, so I think it looks good. 😄

**Review by xnyo — APPROVED** [review]
LGTM, great work!! 👏👏👏

**Comment by andresmgot** [comment]
okay so I believe I know what's happening. 

When using `production` and returning an `response.Error`, the error content is not sent in the HTTP response:

https://github.com/grafana/grafana/blob/main/pkg/api/response/response.go#L232
```go
func Error(status int, message string, err error) *NormalResponse {
...
	if err != nil {
		if setting.Env != setting.Prod {
			data["error"] = err.Error()
		}
	}
```

I don't think we can do anything with that, since is a generic security measure.

Apart from that, when using a valid `OS`, like `Windows` but with a wrong combination with a version, the Grafana download API returns a 400, entering in this case:

https://github.com/grafana/grafana/blob/main/pkg/api/plugins.go#L467
```go
		var clientError repo.Response4xxError
		if errors.As(err, &clientError) {
			return response.Error(clientError.StatusCode, clientError.Message, err)
		}
```

And so, in the title, you get a proper message (even though you don't get the full error message):

Screenshot from 2023-04-14 15-07-36

If I use a wrong OS, like `nope`, then the download API returns a 500 (which is probably not correct, but that's a different story). In that case, the error returned is the default case:

https://github.com/grafana/grafana/blob/main/pkg/api/plugins.go#L473
```
		return response.Error(http.StatusInternalServerError, "Failed to install plugin", err)
```

which is expected since we don't know which kind of error is returning.

So long story short, I don't think we need to take further action but let me know if there is still something I am missing.

**Comment by andresmgot** [comment]
okay, now you should get the proper error even in `production` :pray: thanks for the debugging info!

File: pkg/api/plugins.go [hunk 1]
```diff
@@ -469,6 +469,10 @@ func (hs *HTTPServer) InstallPlugin(c *contextmodel.ReqContext) response.Respons
 		if errors.Is(err, plugins.ErrInstallCorePlugin) {
 			return response.Error(http.StatusForbidden, "Cannot install or change a Core plugin", err)
 		}
+		var archError repo.ErrArchUnsupported
+		if errors.As(err, &archError) {
+			return response.Error(http.StatusConflict, archError.Error(), nil)
```
Comment (wbrowne) [inline pkg/api/plugins.go:474]: Curious about the choice of status conflict. I think of it as not found, as you're asking Grafana to install a plugin, but we could not find a version that supports your architecture. I would expect a conflict if maybe I tried to install a version of a plugin that was already installed?

Comment (andresmgot) [reply pkg/api/plugins.go:474]: I copied it from `versionUnsupportedErr` case above but I agree with the reasoning, it makes sense for this to return a 404 (maybe the `versionUnsupportedErr` is wrongly set?)

Comment (wbrowne) [reply pkg/api/plugins.go]: Ah yes - good for pointing that out. We are currently able to tell the difference between when a version does not exist and when a version exists but does support your system. Maybe then for consistency between the two, this should actually be `ErrArchNotFound`?

Comment (andresmgot) [reply pkg/api/plugins.go]: makes sense :+1:

**Review by wbrowne — APPROVED** [review]
LGTM

**Review by xnyo — APPROVED** [review]
LGTM! Great work :rocket:
